validator/client/src/main/java/tech/pegasys/teku/validator/client/proposerconfig/AbstractProposerConfigProvider.java

```diff
@@ -67,39 +68,33 @@ public synchronized SafeFuture<Optional<ProposerConfig>> getProposerConfig() {
     }
 
     futureProposerConfig =
-        Optional.of(
-            asyncRunner
-                .runAsync(
-                    () -> {
-                      lastProposerConfig = Optional.of(internalGetProposerConfig());
-                      lastProposerConfigTimeStamp = timeProvider.getTimeInSeconds();
-                      return lastProposerConfig;
-                    })
-                .orTimeout(30, TimeUnit.SECONDS)
-                .exceptionally(
-                    throwable -> {
-                      if (lastProposerConfig.isPresent()) {
-                        LOG.warn(
-                            "An error occurred while obtaining config, providing last loaded config",
-                            throwable);
-                        return lastProposerConfig;
-                      }
-                      throw new RuntimeException(
-                          "An error occurred while obtaining config and there is no previously loaded config",
-                          throwable);
-                    })
-                .thenPeek(
-                    proposerConfig ->
-                        LOG.info(
-                            "Proposer config successfully loaded. It contains the default configuration and {} specific configuration(s).",
-                            proposerConfig.orElseThrow().getNumberOfProposerConfigs()))
-                .alwaysRun(
-                    () -> {
-                      synchronized (this) {
-                        futureProposerConfig = Optional.empty();
-                      }
-                    }));
-    return futureProposerConfig.get();
+        asyncRunner
+            .runAsync(this::internalGetProposerConfig)
+            .orTimeout(30, TimeUnit.SECONDS)
+            .thenApply(this::updateProposerConfig)
+            .exceptionally(this::handleException);
+
+    return futureProposerConfig;
+  }
+
+  private synchronized Optional<ProposerConfig> handleException(final Throwable throwable) {
+    if (lastProposerConfig.isPresent()) {
+      LOG.warn("An error occurred while obtaining config, providing last loaded config", throwable);
+      return lastProposerConfig;
+    }
+    throw new RuntimeException(
+        "An error occurred while obtaining config and there is no previously loaded config",
+        throwable);
+  }
```

Comment:
why synchronized? and do we have tests showing why the runtimeException should be there? it seems like this might be a problem...

Reply:
> why synchronized? 

we read `lastProposerConfig` there (I want to make sure the thread reads the last value) and I become paranoid due to the unexplained bug :)

Reply:
> runtimeException should be there?

I'm ok with that, because we cannot even start the node if we never loaded a config. So having a valid `lastProposerConfig` present is a must.

Reply:
`ProposerConfigManager::initialize` wants it to fail exceptionally to prevent node to startup